### PR TITLE
Input area interaction

### LINKED ISSUE #65: Improve interaction model of input area on Copy challenges
The component currently uses the [react-individual-character-input-boxes](https://www.npmjs.com/package/react-individual-character-input-boxes) module. The expected behavior is that of a regular text box, so it feels a bit off when trying to do anything more than a simple paste. 

Let's try a simple text box approach that is styled to still show the underlines (maybe another hidden text area behind the input that intelligently shows the underscored for blank characters).

This affects the two copy challenge pages we have:
- [Genesis](https://savingsatoshi.com/chapters/chapter-1/genesis)
- [Transacting](https://savingsatoshi.com/chapters/chapter-1/transacting/transacting-1)

Screenshot:
image

### PR DESCRIPTION
**Pull request by benalleng**
Adds better userInput with standard textarea styled to mimic the react individual character input boxes 'RICIBs'

Resolves #65

## Pull Request checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build was run locally and any changes were pushed
- [x] Lint has passed locally and any fixes were made for failures

## Pull Request type

Please check the type of change your PR introduces:

- [ ] :bug: Bugfix
- [x] :sparkles: Feature
- [x] :lipstick: :art: Code style update (formatting, renaming)
- [ ] :recycle: Refactoring (no functional changes, no api changes)
- [ ] :zap: Performance
- [ ] :pencil: Documentation content changes
- [ ] :wrench: Other (please describe):

## What is the current behavior?

RICIBs gives a lot of footguns to the user to type the answers incorrectly. While this is not a problem in Chapter 1 as the user is only required to paste it could become a problem in higher challenges.

## What is the new behavior?

- Typing and Paste-ing behave like a standard texarea

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

I think it is worth considering utilizing RICIBs or perhaps another custom text area when the user is building their own transactions. The data is broken into chunks and RICIBs are good for allowing the user to type it in however the please as loosely demonstrated below. I still belive in most cases, the new input system is better in most cases

Before:

image

After: 

image
image

### CODE REVIEW

**Comment by benalleng** [comment]
I would like to know if we prefer to still have the underscores disappear? just to make it as similar to the previous input field as possible.

**Comment by satsdev** [comment]
Tested on desktop and mobile (IOS). Nicely improved usability on desktop. I did not see any issues besides what you already have listed in TODO.

**Comment by GBKS** [comment]
Thank you a ton for tackling this one, several people have complained about the interaction model of this component. I also realized that I never provided full design specs for this component, so I put something together here (and [in Figma](https://www.figma.com/file/LqjK3Tpvd9KJ4buFArCJBQ/Saving-Satoshi?node-id=1393%3A15989&t=fF6Oe8NLCx1yFm5D-1)).
image

To your question, I think it's much easier on the eyes when the underscores disappear. It might also be best if we use color to indicate incorrect input (blue) and valid input (green). I prefer blue over red, because it's not really an error (red is always so serious and makes you feel bad).

For mobile, I reduced the font size to 18px (from 30px), line height is at 180% at letter-spacing at 5%. The green and blue should use our standardized theme CSS color variables (they may slightly differ for each theme).

Hope that's useful, let me know how it looks.

File: components/UserInput.tsx
Comment (GBKS) [inline components/UserInput.tsx:24]: What do you need the line breaks for?

File: components/UserInput.tsx
Comment (GBKS) [inline components/UserInput.tsx:23]: Would be amazing if you could use Tailwind instead of styles.

File: components/UserInput.tsx
Comment (GBKS) [inline components/UserInput.tsx:31]: We may need these to be customizable. Do you know if browser strictly enforce rows and cols or if they can be overwritten with CSS?

File: components/UserInput.tsx
Comment (GBKS) [inline components/UserInput.tsx:52]: Width should be adjustable per page (maybe via a max-width), it currently causes a horizontal scroll-bar. Ideally, we have roughly equal width rows (e.g. we have 150 characters = 3 rows of 50 on desktop and 5 rows of 30 on mobile) for a consistent look, but I am not sure if we can easily get full control over this and can use max-width as a shortcut.

File: components/UserInput.tsx
Comment (GBKS) [inline components/UserInput.tsx:60]: You may not need `position:relative` and the `z-index` here if you use `pointer-events: none`, which makes it non-selectable.

File: components/UserInput.tsx
Comment (GBKS) [inline components/UserInput.tsx:30]: Do we need `id` and `name` attributes?

File: components/UserInput.tsx
Comment (GBKS) [inline components/UserInput.tsx:6]: Don't think you need this import if you use the `font-space-mono` Tailwind class.

**Review by GBKS — CHANGES_REQUESTED** [review]
Thanks again for tackling this. I reviewed the code and had a few comments. Some around consistency with the rest of the code, and some about making this component more flexible for various inputs and screen dimensions. Let me know if you have questions.

File: components/UserInput.tsx
Comment (benalleng) [reply components/UserInput.tsx]: not needed

**Comment by benalleng** [comment]
>     * Weird offset when removing parts of the last line
>       image

This is currently one of my main blockers. While i can set the test-align-last: center for the underscores I am unsure how to achieve this with a textarea or any kind of editable input. Perhaps this is where Christoph mentioning changing it to be a content editable div would give me better functionality.

**Comment by ecurrencyhodler** [comment]
Another issue is that after pasting in the hex, I am unable to move onto the next chapter.
test

**Comment by ecurrencyhodler** [comment]
I tested this on mobile. Seems like the sizing is off as well. Not sure if we want to address that in this PR. 

But like the updates I see on the web!
IMG_1334

**Comment by afterburn** [comment]
Hey, just sharing my thoughts on this. I think the approach you've chosen is promising. I hope you can resolve the remaining issues using this approach, but if that turns out to be rather difficult we might need to consider using `contenteditable`. In my opinion, the problem with regular inputs and textareas is that they do not provide fine-grained control over individual characters/tokens, which is what we seem to be trying to achieve here.

I'm sure you'll be able to get different colors working for correct/incorrect characters, and get the underscores lined up correctly, but I just wanted to share this [code](https://jsfiddle.net/afterburner/rejbty21/1/) I had laying around from some old, forgotten project of mine, which I modified to more closely resemble the input we're trying to create here. While it's by no means ready for production, it uses a similar approach (overlaying two elements) but I think it provides more control over individual tokens. I don't know, perhaps it could be useful to you.

On the flip side, it's also worth mentioning that `contenteditable` is often way more involved (as you can see from the code example) and it's very prone to bugs, so it might not be worth the effort. I think if we can avoid using `contenteditable` and still meet requirements that would be best.

**Comment by benalleng** [comment]
@ecurrencyhodler Hey I'm having some trouble with the reproduceability of your issues, perhaps you have the settings of an older preview cached? 

image


image

**Comment by benalleng** [comment]
@afterburn I think you're right on the contenteditable div being a better approach afterall, and I appreciate the code you provided!  I had actually begun to refactor to contenteditable div already I just feel that it isn't quite ready to be reviewed. some of your code will certainly be helpful! I think unfortunately creating using a textarea is causing the last couple of blockers to match the figma GBKS posted

**Comment by benalleng** [comment]
I played around with the fiddle a little bit today, got it right here now i just have to refactor it for react :sweat_smile: 
https://jsfiddle.net/0nb8g7mc/

**Comment by GBKS** [comment]
@benalleng is it possible to restrict the characters a user can input to avoid some of those edge cases (and save you a headache)? Seems like it [should be](https://gist.github.com/hygull/356d1c5f043b1564d9fb59e58a490272).

**Comment by afterburn** [comment]
Hey @benalleng, you should be able to prevent certain characters from being input on the keypress event. You can see me doing the same in my fiddle. 😁

**Comment by ecurrencyhodler** [comment]
@benalleng you may need to rebase this with https://github.com/saving-satoshi/saving-satoshi/pull/118 now merged.

**Comment by benalleng** [comment]
@GBKS, @ecurrencyhodler  or @afterburn I have rebased this following #118 and I think its ready for y'all to take a look for review!

**Comment by ecurrencyhodler** [comment]
Hmmm not that I can think of. I just copied and pasted the hex into prod and it worked. Which hex are you using?
Screenshot 2023-02-01 at 10 10 37 AM

File: content/lessons/genesis/genesis.tsx
Comment (benalleng) [inline content/lessons/genesis/genesis.tsx:17]: I am just grabbing it from the answer posted in the genesis content page, I am going to try and make catch to make sure that only numbers/characters can be pasted in correctly

**Comment by benalleng** [comment]
@ecurrencyhodler I added a regex like function that should remove spaces and line breaks to prevent any invisible characters when pasting

**Comment by ecurrencyhodler** [comment]
tACK! Looks good to me. Great job!
Screenshot 2023-02-01 at 9 21 13 PM

File: ui/common/Input.tsx
Comment (afterburn) [inline ui/common/Input.tsx:23]: Nitpick but you can chain methods on the String class:
```js
const underscores = '_'.repeat(answer.length).substring(0, answer.length - textAreaValue.length)
```
Alternatively you can do:
```js
const underscores = '_'.repeat(Math.max(0, answer.length - textAreaValue.length))
    
```

File: ui/common/Input.tsx
Comment (afterburn) [inline ui/common/Input.tsx:13]: Nitpick but let's try to steer away from abbreviations like ans for answer, these might be confusing for non-english folk and could leave some room for interpretation, which we want to avoid.

File: ui/common/Input.tsx
Comment (afterburn) [inline ui/common/Input.tsx:28]: Another nitpick but let's always use curly braces, even for switches. Though its not incorrect to do it this way in this specific scenario I think curly braces help prevent issues with scoping, so it's a good practice to have.

File: ui/common/Input.tsx
Comment (afterburn) [inline ui/common/Input.tsx:62]: Let's try to keep the jsx markup separated from the logic, this helps with readability. So what I mean by that is simply create a handler function below `handleKeyDown`, maybe call it `handleChange` or something akin to that. Also makes things easier to find.

Comment (afterburn) [inline ui/common/Input.tsx:62]: I'm not sure what this code is doing:
```js
e.target.value.replace(/[\s\u00A0]+/g, '').replace(/\r?\n|\r/g, '').slice(0, answer.length).toLowerCase()
```

Its use seems to be to remove linebreaks from the users input, if that is the case I think a better approach is to use the `keypress` event and check what keyCode is being pressed. If the keyCode matches certain characters you wish to block you can `return false`. This way you dont have to modify the value after the fact.

**Comment by afterburn** [comment]
Hey Ben thank you for all the hard work so far! I left some nitpicky comments but I hope they are helpful.

I also noticed that I'm able to input characters other than `a-z0-9` (see screenshot). Not sure if we want that for this specific lesson but @GBKS can confirm.
image

Perhaps out of scope for this PR but I think it makes sense to have the `<LessonPrompt>` component take in a `pattern` prop, which could be a regex (or a function so you we can create custom validators), and in the `<LessonPrompt>` component you can validate the input against that regex (or function) to make sure no invalid characters can be input. Something like this:
```jsx
<LessonPrompt
  className="max-w-[840px] px-4 py-8 md:items-center"
  label="Enter the OP_RETURN type"
  answer={answer}
  pattern={/^[a-z0-9]+$/i}
  onChange={setUserInput}
/>
```

File: ui/common/Input.tsx
Comment (benalleng) [reply ui/common/Input.tsx:62]: This was intended to handle formatting when pasting in text, it is intended to catch linebreaks and whitespace characters, i refactored it to make it more clear what does what

Comment (afterburn) [reply ui/common/Input.tsx]: Okay I see, let's use [the paste event](https://developer.mozilla.org/en-US/docs/Web/API/Element/paste_event) for that though. Something like this:

```js
function handlePaste(event) {
  event.preventDefault()

  const pasteData = (event.clipboardData || window.clipboardData).getData('text')
    .replace(/\s+/g, '')

  document.execCommand('insertText', false, pasteData)
}
```

```jsx
<textarea
  onPaste={handlePaste}
  ...
/>
```

**Comment by benalleng** [comment]
great idea about the pattern matching on the UI!

File: ui/common/Input.tsx
Comment (benalleng) [reply ui/common/Input.tsx]: wow well that does it! thanks for showing me this solution!

Comment (benalleng) [reply ui/common/Input.tsx]: I tried implementing this paste event but it no longer set the textarea to the clipboard remaining blank after pasting

Comment (afterburn) [reply ui/common/Input.tsx]: Hey @benalleng, are you referring to this code or have you not pushed your attempt at this to GitHub?

```jsx
  const handlePaste = (event) => {
    const formattedText = event
      .replace(/[\s\u00A0]+/g, '')
      .replace(/\r?\n|\r/g, '')
      .slice(0, answer.length)
      .toLowerCase()
    return formattedText
  }
  
  const handleChange = (event) => {
    let formattedText = handlePaste(event.target.value)
    ...
  })
  ```
  
  ```jsx
  <textarea
    onChange={handleChange}
    ...
  />
  ```

**Comment by afterburn** [comment]
I gave this another test but I'm still able to enter characters other than `a-z0-9`. Let's make sure it doesn't accept any other characters than those and we can leave the pattern prop for another PR unless you want to resolve it in this one.

**Comment by benalleng** [comment]
@afterburn Sorry it wasn't ready for you to take a look at it from the call, I had reverted to a prior commit. This afternoon I took another pass and I got it working! The `<LessonPrompt />` component now accepts an optional pattern to allow the route template code to limit what is input. upon pasting the `onPaste` handler gets rid of anything that doesn't match the regex expression.  The `onChange` handler has also been simplified separating duties somewhat

File: ui/common/Input.tsx
Comment (benalleng) [reply ui/common/Input.tsx]: I resolved this function with a slight adjustment to the onPaste handler

File: ui/common/Input.tsx
Comment (afterburn) [inline ui/common/Input.tsx:66]: Some strange formatting here, have you ran eslint/prettier?

**Review by afterburn — CHANGES_REQUESTED** [review]
Looks good, couldn't find any bugs. Left a small comments and you have some conflicts but after those have been resolved LGTM.

File: app/[lang]/chapters/[slug]/page.tsx
Comment (afterburn) [inline app/[lang]/chapters/[slug]/page.tsx:2]: You can combine these into a single import statement
```jsx
import { chapters, introductions } from 'content'
```

Comment (afterburn) [reply app/[lang]/chapters/[slug]/page.tsx:2]: Oh I actually think merge conflicts caused these, nevermind this comment.

**Comment by benalleng** [comment]
There are a few new conflicts introduced by the new template layouts, I am going to convert this to a draft until it is fully merged with master.

**Comment by benalleng** [comment]
@afterburn Should be good to go with the new challenge templates but lets have one final review just to check!

**Review by afterburn — APPROVED** [review]

**Comment by GBKS** [comment]
Congrats @benalleng for getting this massive update merged, appreciate your work on this.

Sorry to report a bug right away, I'm finding that when I start entering text in the input field on Safari (desktop and mobile), it instantly switches from multi line to single line, stretching beyond the browser frame. This might just be a simple CSS setting (max-width or so).

**Pull request merged**
